src/pcs/kzg/accumulation.rs

```diff
@@ -0,0 +1,196 @@
+        proof: &Self::Proof,
+    ) -> Result<PCS::Accumulator, Error> {
+        let (lhs, rhs) = instances
+            .iter()
+            .cloned()
+            .map(|accumulator| (accumulator.lhs, accumulator.rhs))
+            .chain(proof.blind.clone())
+            .unzip::<_, _, Vec<_>, Vec<_>>();
+
+        let powers_of_r = proof.r.powers(lhs.len());
+        let [lhs, rhs] = [lhs, rhs].map(|msms| {
+            msms.into_iter()
+                .zip(powers_of_r.iter())
+                .map(|(msm, r)| Msm::<C, L>::base(msm) * r)
+                .sum::<Msm<_, _>>()
+                .evaluate(None)
+        });
+
+        Ok(KzgAccumulator::new(lhs, rhs))
+    }
+}
+
+#[derive(Clone, Copy, Debug, Default)]
+pub struct KzgAsProvingKey<C>(pub Option<(C, C)>);
+
+impl<C: Clone> KzgAsProvingKey<C> {
+    pub fn new(g: Option<(C, C)>) -> Self {
+        Self(g)
+    }
+
+    pub fn zk(&self) -> bool {
+        self.0.is_some()
+    }
+
+    pub fn vk(&self) -> KzgAsVerifyingKey {
+        KzgAsVerifyingKey(self.zk())
+    }
+}
+
+#[derive(Clone, Copy, Debug, Default)]
+pub struct KzgAsVerifyingKey(bool);
+
+impl KzgAsVerifyingKey {
+    pub fn zk(&self) -> bool {
+        self.0
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct KzgAsProof<C, L, PCS>
+where
+    C: CurveAffine,
+    L: Loader<C>,
+    PCS: PolynomialCommitmentScheme<C, L, Accumulator = KzgAccumulator<C, L>>,
+{
+    blind: Option<(L::LoadedEcPoint, L::LoadedEcPoint)>,
+    r: L::LoadedScalar,
+    _marker: PhantomData<PCS>,
+}
+
+impl<C, L, PCS> KzgAsProof<C, L, PCS>
+where
+    C: CurveAffine,
+    L: Loader<C>,
+    PCS: PolynomialCommitmentScheme<C, L, Accumulator = KzgAccumulator<C, L>>,
+{
+    fn read<T>(
+        vk: &KzgAsVerifyingKey,
+        instances: &[PCS::Accumulator],
+        transcript: &mut T,
+    ) -> Result<Self, Error>
+    where
+        T: TranscriptRead<C, L>,
+    {
+        assert!(instances.len() > 1);
+
+        for accumulator in instances {
+            transcript.common_ec_point(&accumulator.lhs)?;
+            transcript.common_ec_point(&accumulator.rhs)?;
+        }
+
+        let blind = vk
+            .zk()
+            .then(|| Ok((transcript.read_ec_point()?, transcript.read_ec_point()?)))
+            .transpose()?;
+
+        let r = transcript.squeeze_challenge();
+
+        Ok(Self {
+            blind,
+            r,
+            _marker: PhantomData,
+        })
+    }
+}
+
+impl<C, PCS> AccumulationSchemeProver<C, PCS> for KzgAs<PCS>
+where
+    C: CurveAffine,
+    PCS: PolynomialCommitmentScheme<C, NativeLoader, Accumulator = KzgAccumulator<C, NativeLoader>>,
+{
+    type ProvingKey = KzgAsProvingKey<C>;
+
+    fn create_proof<T, R>(
+        pk: &Self::ProvingKey,
+        instances: &[PCS::Accumulator],
+        transcript: &mut T,
+        rng: R,
+    ) -> Result<PCS::Accumulator, Error>
+    where
+        T: TranscriptWrite<C>,
+        R: Rng,
+    {
+        assert!(instances.len() > 1);
```

Comment:
@han0110 I'm hitting this one. Shouldn't that be `instances.len > 0`?

Reply:
Make sense, updated in https://github.com/privacy-scaling-explorations/plonk-verifier/pull/5/commits/5a5a6478525817aaaf05b305871cfca7b0704261.